pygfx/renderers/wgpu/_wgpurenderer.py

```diff
@@ -1039,6 +1041,14 @@ def _update_sampler(self, resource):
         )
         resource._wgpu_sampler = resource.rev, sampler
 
+    def _get_shader_module(self, key, source):
+        """Compile a shader module object, or re-use it from the cache."""
+        # todo: make this work for objects following the ShaderSourceTemplate interface
+        if key not in self._shader_cache:
+            m = self._device.create_shader_module(code=source)
+            self._shader_cache[key] = m
+        return self._shader_cache[key]
```

Comment:
just wondering if you also need to implement code to release the resources?

Reply:
mmm, yeah, in practice the number of shaders used is probably finite, but it does not have to be.